src/mount/fuse/sfs_meta_fuse.cc

```diff
@@ -639,126 +657,124 @@ void sfs_meta_open(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi) {
 	} else {
-		pathinfo = new pathbuf();
-
-		pathinfo->changed = 0;
-		pathinfo->size = strlen((char*)path) + 1;
-		if (!(pathinfo->p = (char*)malloc(pathinfo->size))) {
-			safs_pretty_syslog(LOG_EMERG, "out of memory");
-			delete pathinfo;
-			return;
-		}
-		memcpy(pathinfo->p, path, pathinfo->size - 1);
-		pathinfo->p[pathinfo->size - 1] = '\n';
+		pathinfo->changed = false;
+		pathinfo->size = strlen((char *)path) + 1;
+		pathinfo->path = std::string((char *)path);
+		pathinfo->path.append("\n"); 
 		fi->direct_io = 1;
-		fi->fh = (unsigned long)pathinfo;
+		fi->fh = reinterpret_cast<uintptr_t>(pathinfo.release());
 
-		if (fuse_reply_open(req, fi) == -ENOENT) {
-			fi->fh = 0;
-			free(pathinfo->p);
-			delete pathinfo;
-		}
+		if (fuse_reply_open(req, fi) == -ENOENT) { fi->fh = 0; }
 	}
 }
 
 void sfs_meta_release(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi) {
-	if (ino==SPECIAL_INODE_MASTERINFO) {
-		fuse_reply_err(req,0);
+	if (ino == SPECIAL_INODE_MASTERINFO) {
+		fuse_reply_err(req, 0);
 		return;
 	}
-	pathbuf *pathinfo = (pathbuf *)((unsigned long)(fi->fh));
+
+	auto pathinfo = std::unique_ptr<PathBuffer>(reinterpret_cast<PathBuffer *>(fi->fh));
+
 	std::unique_lock lock(pathinfo->lock);
 	if (pathinfo->changed) {
-		if (pathinfo->p[pathinfo->size-1]=='\n') {
-			pathinfo->p[pathinfo->size-1]=0;
+		if (pathinfo->path.back() == '\n') {
+			pathinfo->path.pop_back();
 		} else {
-			pathinfo->p = (char*) realloc(pathinfo->p,pathinfo->size+1);
-			pathinfo->p[pathinfo->size]=0;
+			pathinfo->path.resize(pathinfo->size + 1);
+			pathinfo->path.back() = '\0';
 		}
-		fs_settrashpath(ino,(uint8_t*)pathinfo->p);
+		fs_settrashpath(ino, reinterpret_cast<const uint8_t *>(pathinfo->path.c_str()));
 	}
 	lock.unlock();
-	free(pathinfo->p);
-	delete pathinfo;
+
 	fi->fh = 0;
-	fuse_reply_err(req,0);
+	fuse_reply_err(req, 0);
 }
 
-void sfs_meta_read(fuse_req_t req, fuse_ino_t ino, size_t size, off_t off, struct fuse_file_info *fi) {
-	pathbuf *pathinfo = (pathbuf *)((unsigned long)(fi->fh));
-	if (ino==SPECIAL_INODE_MASTERINFO) {
-		uint8_t masterinfo[14];
-		fs_getmasterlocation(masterinfo);
-		masterproxy_getlocation(masterinfo);
+void sfs_meta_read(fuse_req_t req, fuse_ino_t ino, size_t size, off_t off,
+                   struct fuse_file_info *fi) {
 #ifdef MASTERINFO_WITH_VERSION
-		if (off>=14) {
-			fuse_reply_buf(req,NULL,0);
-		} else if (off+size>14) {
-			fuse_reply_buf(req,(char*)(masterinfo+off),14-off);
+	constexpr off_t kMasterInfoSize =
+	    14;  // size of the master info buffer with version information.
 #else
-		if (off>=10) {
-			fuse_reply_buf(req,NULL,0);
-		} else if (off+size>10) {
-			fuse_reply_buf(req,(char*)(masterinfo+off),10-off);
+	constexpr off_t kMasterInfoSize =
+	    10;  // size of the master info buffer without version information.
 #endif
+	auto pathinfo = reinterpret_cast<PathBuffer *>(fi->fh);
+
+	if (ino == SPECIAL_INODE_MASTERINFO) {
+		uint8_t masterinfo[14];
```

Comment:
If only one size is supported (see comment above), then,
```suggestion
		uint8_t masterinfo[kMasterInfoSize];
```

Reply:
Good, I will take this as a consideration to remove the few references to all master info size in another PR. To make it clearer when deprecating it.